File: research/spotvol/black_scholes.py
import math

def _norm_cdf(x: float) -> float:
    return 0.5 * (1 + math.erf(x / math.sqrt(2)))

# European option using Black-Scholes model
def black_scholes_price(S: float, K: float, T: float, r: float, sigma: float, is_call: bool) -> float:
    # z-scores measuring distance from strike 
    d1 = (math.log(S / K) + (r + 0.5 * sigma**2) * T) / (sigma * math.sqrt(T))
    d2 = d1 - sigma * math.sqrt(T)
    if is_call:
        return S * _norm_cdf(d1) - K * math.exp(-r * T) * _norm_cdf(d2)
    else:
        # put
        return K * math.exp(-r * T) * _norm_cdf(-d2) - S * _norm_cdf(-d1)

def implied_vol(market_price: float, S: float, K: float, T: float, r: float, is_call: bool) -> float:
    if T <= 0 or market_price <= 0:
        return math.nan

    # option worth right now if exercised
    intrinsic = max(S - K * math.exp(-r * T), 0) if is_call else max(K * math.exp(-r * T) - S, 0)

    # bad data 
    if market_price < intrinsic:
        return math.nan

    # bisection for vol
    lo, hi = 1e-6, 5.0
    for _ in range(100):
        mid = (lo + hi) / 2
        if black_scholes_price(S, K, T, r, mid, is_call) < market_price:
            lo = mid
        else:
            hi = mid

    if hi - lo < 1e-8:
        return mid
    return math.nan

File: src/black_scholes.cpp
#include "spotvol/black_scholes.hpp"

#include <algorithm>
#include <cmath>
#include <limits>

namespace spotvol {

static double normCdf(double x) {
    return 0.5 * (1.0 + std::erf(x / std::sqrt(2.0)));
}

double bsPrice(double S, double K, double T, double r, double sigma, bool isCall) {
    double sqrtT = std::sqrt(T);
    double d1 = (std::log(S / K) + (r + 0.5 * sigma * sigma) * T) / (sigma * sqrtT);
    double d2 = d1 - sigma * sqrtT;
    if (isCall) {
        return S * normCdf(d1) - K * std::exp(-r * T) * normCdf(d2);
    }
    return K * std::exp(-r * T) * normCdf(-d2) - S * normCdf(-d1);
}

double impliedVol(double marketPrice, double S, double K, double T, double r, bool isCall) {
    if (T <= 0.0 || marketPrice <= 0.0) {
        return std::numeric_limits<double>::quiet_NaN();
    }

    double disc = std::exp(-r * T);
    double intrinsic = isCall
        ? std::max(S - K * disc, 0.0)
        : std::max(K * disc - S, 0.0);

    if (marketPrice < intrinsic) {
        return std::numeric_limits<double>::quiet_NaN();
    }

    double lo = 1e-6, hi = 5.0;
    for (int i = 0; i < 100; ++i) {
        double mid = 0.5 * (lo + hi);
        if (bsPrice(S, K, T, r, mid, isCall) < marketPrice) {
            lo = mid;
        } else {
            hi = mid;
        }
    }

    if (hi - lo < 1e-8) {
        return 0.5 * (lo + hi);
    }
    return std::numeric_limits<double>::quiet_NaN();
}

}
